students/DanCornutt/session06/mailroom.py

```diff
@@ -0,0 +1,214 @@
+            raise ValueError
+    except ValueError:
+        print("Donation Error, numbers only. Value must be number greater than 0.01")
+        return None
+    else:
+        return float(donation)
+
+
+def thank_you_all():
+    """Writes thank you email to all users"""
+    thank_you(all_users=True)
+
+
+def report():
+    """Finds column widths, sorts donors based on amount donated.
+    :returns: report string
+    """
+    rpt_sheet = []
+    len_col = OrderedDict({
+        "n_size": len("Donor Name"),
+        "t_size" : len("Total Given"),
+        "ng_size": len("Num Gifts"),
+        "ag_size": len("Average Gift"),
+        "nm": "Donor Name",
+        "tot": "Total Given",
+        "ng": "Num Gifts",
+        "ag": "Average Gift"
+        })
+
+    for d in DONORS.items():
+        #finds max column sizes
+        if len(d[0]) > len_col["n_size"]:
+            len_col["n_size"] = len(d[0])
+        if len(str(sum(d[1]))) > len_col["t_size"]:
+            len_col["t_size"] = len(str(sum(d[1])))
+        if len((d[1])) > len_col["ng_size"]:
+            len_col["ng_size"] = len((d[1]))
+        if len(str(sum(d[1])/len(d[1]))) > len_col["ag_size"]:
+            len_col["ag_size"] = len(str(sum(d[1])/len(d[1])))
+        #creates report data
+        rpt_sheet.append((d[0], sum(d[1]), len(d[1]), sum(d[1])/len(d[1])))
+    rpt_sheet.sort(key=return_total, reverse=True)
+
+    #compiles data into wanted format
+    #header
+    sheet = (
+        "{nm:{n_size}} | {tot:{t_size}} | {ng:{ng_size}} |{ag:<{ag_size}}\n".format(
+            **len_col) + ("-" * sum(list(len_col.values())[:4]))
+    )
+    #data
+    for d in rpt_sheet:
+        sheet = sheet + (
+            "\n{:{n_size}} |${:>{t_size},.2f} | {:>{ng_size}} |$ {:<{ag_size},.2f}"
+            .format(*d, **len_col)
+        )
+    return sheet
+
+
+def make_report():
+    """Prints report"""
+    info = report()
+    print(info)
+
+def return_total(elem):
+    """sorting function for list"""
+    return elem[1]
+
+
+def donor_list():
+    """prints donor in dict"""
+    for k in DONORS:
+        print(k)
+
+
+def unknown():
+    """Handles unknown user input"""
+    print("That is not a valid response!")
+
+
+def quit_menu():
+    """Quits menu, returns 'exit menu'. """
+    print("Quitting this menu now.")
+    return "exit menu"
+
+
+def menu_selection(prompt, dispatch_dict):
+    """dispatch function for mailroom"""
+    while True: #this loops forever, until quit is selected
+        response = input(prompt).strip()
+        response = response[:1].lower()
+        if dispatch_dict.get(response, unknown)() == "exit menu":
+            break
+
+MAIN_DISPATCH = {
+    "1": thank_you,
+    "2": make_report,
+    "3": thank_you_all,
+    "4": donor_db,
+    "5": quit_menu,
+    }
+
+DATABASE_DISPATCH = {
+    "1": donor_list,
+    "2": edit_donor,
+    "3": quit_menu
+}
+
+if __name__ == "__main__":
+    menu_selection(WELCOME_PROMPT, MAIN_DISPATCH)
```

Comment:
this looks pretty much the same -- I assume a copy for testing next lesson.

But note: you still need to clean up the print() and input() a bit more -- they should be isolated in functions that do almost nothing else, so you can test as much as possible.